core/src/main/scala/org/apache/spark/sql/internal/ColumnTableBulkOps.scala

```diff
@@ -69,42 +54,60 @@ object ColumnTableBulkOps {
         val condition = prepareCondition(sparkSession, table, subQuery, putKeys.get)
 
         val keyColumns = getKeyColumns(table)
-        val updateSubQuery = Join(table, subQuery, Inner, condition)
+        var updateSubQuery: LogicalPlan = Join(table, subQuery, Inner, condition)
         val updateColumns = table.output.filterNot(a => keyColumns.contains(a.name))
+        val updateExpressions = updateSubQuery.output.takeRight(updateColumns.length)
 
-        val cacheSize = Property.PutIntoInnerJoinCacheSize
-            .getOption(sparkSession.sparkContext.conf) match {
-          case Some(size) => size.toInt
-          case None => Property.PutIntoInnerJoinCacheSize.defaultValue.get
-        }
+        val cacheSize = ExternalStoreUtils.sizeAsBytes(
+          Property.PutIntoInnerJoinCacheSize.get(sparkSession.sqlContext.conf),
+          Property.PutIntoInnerJoinCacheSize.name, -1, Long.MaxValue)
+
+        val updatePlan = Update(table, updateSubQuery, Seq.empty,
+          updateColumns, updateExpressions)
+        val updateDS = new Dataset(sparkSession, updatePlan, RowEncoder(updatePlan.schema))
+        val analyzedUpdate = updateDS.queryExecution.analyzed.asInstanceOf[Update]
+        updateSubQuery = analyzedUpdate.child
 
-        val doInsertJoin = if (updateSubQuery.statistics.sizeInBytes <= cacheSize) {
+        val doInsertJoin = if (subQuery.statistics.sizeInBytes <= cacheSize) {
           val joinDS = new Dataset(sparkSession,
             updateSubQuery, RowEncoder(updateSubQuery.schema))
+          if (sparkSession.sharedState.cacheManager
+              .lookupCachedData(updateSubQuery).isDefined){
+            joinDS.unpersist(blocking = true)
+          }
```

Comment:
This means every session will retain a cached result, so won't this keep growing unncessarily? What was the problem with earlier one which used to uncache at the end of executeCollect()? Perhaps what is required there is equivalent of "blocking = true" there.

Reply:
This is workaround. Before I was putting it in contextObject and was clearing after execute all.But looks like contextObject itself was getting cleared. Hence the cached data remained as it is.

Reply:
Ok, but this will keep growing especially if updateSubQuery is different each time (which will normally be the case). I think you can have the key "CACHED_PUTINTO_UPDATE_PLAN" define in SnappySession object, and do the cleanup in SnappySession.clearContext. Later we can think of how to make this generic.